src/Plugin/formatter/FormatterJsonApi.php

```diff
@@ -309,6 +320,142 @@ protected function moveEmbedsToIncluded(array &$output, array $included) {
         $include_keys[] = $include_key;
       }
     }
+    return $output;
+  }
+
+  /**
+   * Remove the unnecessary fields from the response.
+   *
+   * @param string[] $limit_fields
+   *   The list of fields.
+   * @param array $output
+   *   The prepared output.
+   *
+   * @return array
+   *   The filtered contents.
+   */
+  public function limitFields($limit_fields, array $output) {
+    if (!$limit_fields) {
+      return $output;
+    }
+    if (empty($output['__fields'])) {
```

Comment:
matter of taste - for me duplication of method is more extreme then `!empty()` - but I can go either way